Provide tutorial on how to create certificate application pairing via Compass (#550)
kyma-incubator/compass: docs/connector/08-01-establish-secure-connection-with-compass.md

# Establish a secure connection with Compass

To establish a secure connection with Compass and generate the client certificate, follow this tutorial. 

## Prerequisites

- [OpenSSL toolkit](https://www.openssl.org/docs/man1.0.2/apps/openssl.html) to create a Certificate Signing Request (CSR), keys, and certificates which meet high security standards
- Compass (version 1.8 or higher)
- Registered Application
- Runtime connected to Compass

## Steps

1. Get the Connector URL and the one-time token.

    To get the Connector URL and the one-time token which allow you to fetch the required configuration details, use the Compass Console.
    

>     Alternatively, make a call to the Director including the `Tenant` header with Tenant ID and `authorization` header with the Dex Bearer token. Use the following mutation: 

Reviewer: We should provide a description of how to get Dex token. In the current state, it is not the most pleasant way to fetch it but nonetheless - there should be a guide for that.

    
    ```graphql
    mutation { 
        result: generateOneTimeTokenForApplication(id: "{APPLICATION_ID}") { 
            token 
            connectorURL 
        }
    }
    ```
   
   > **NOTE:** The one-time token expires after 5 minutes.

2. Get the CSR information and configuration details from Kyma using the one-time token.

>     To get the CSR information and configuration details, send this GraphQL query with the one-time token included in the `connector-token` header to the Connector URL:

Reviewer: We should emphasize the need of having a `connector-token` header in the call, it is easy to overlook that in the current form.

    ```graphql
    query {
        result: configuration {
            token {
                token
            }
            certificateSigningRequestInfo {
                subject
                keyAlgorithm
            }
            managementPlaneInfo {
                directorURL
                certificateSecuredConnectorURL
            }
        }
    }
    ``` 

    A successful call returns the data requested in the query including a new one-time token.

3. Generate a key and a Certificate Signing Request (CSR).

    Generate a CSR with this command using the certificate subject data obtained with the CSR information: 
    
    ```bash

>     openssl genrsa -out generated.key 2048

Reviewer: The recommended key length is 4096, we should also indicate that it is configurable. Perhaps using {KEY_LENGTH} value set to 4096 is a good idea?

    openssl req -new -sha256 -out generated.csr -key generated.key -subj "{SUBJECT}"

>     openssl base64 -in generated.csr

Reviewer: `generated.*` filenames are not the best ones, I would suggest changing that to `compass-app.*` or something else, which would be less-generic.

Reviewer: Also, there is no information on where the subject was returned in the previous steps.

    ```

4. Sign the CSR and get a client certificate. 

    To get the CSR signed, use the encoded CSR in this GraphQL mutation:
    
    ```graphql
    mutation {

>         result: signCertificateSigningRequest(csr: "{BASE64_ENCODED_CSR}") {

Reviewer: There is no link between `{BASE64_ENCODED_CSR}` and previous sentence, we need to guide the User that this value is the result of `base64 compass-app.csr` command.

            certificateChain
            caCertificate
            clientCertificate
        }
    }
    ```
   

>     Send the modified GraphQL mutation to the Connector URL including the one-time token fetched with the configuration in the `connector-token` header.

Reviewer: Same thing with the header, we need to put emphasis on that.

    The response contains a certificate chain, a valid client certificate signed by the Kyma Certificate Authority (CA), and the CA certificate.
    

>     After you receive the certificates, decode the certificate chain with the base64 method and use it in your application. 

Reviewer: Please include command on how it can be decoded, this is a simple one but worth having here explicitly.

> ---

Reviewer 2: The convention is to use H1 for titles of docs kept on Github, and matadata for docs that are displayed on website. Let's discuss this.

title: Establish a secure connection with Compass
type: Tutorials
---

To establish a secure connection with Compass and generate the client certificate, follow this tutorial. 

## Prerequisites

- [OpenSSL toolkit](https://www.openssl.org/docs/man1.0.2/apps/openssl.html) to create a Certificate Signing Request (CSR), keys, and certificates which meet high security standards

> - Compass (version >= 1.8)

Reviewer 2: ```suggestion
- Compass (version 1.8 or higher)
```

                certificateSecuredConnectorURL
            }
        }
    }
    ``` 

    A successful call returns the data requested in the query including a new one-time token.

3. Generate a key and a Certificate Signing Request (CSR).

    Generate a CSR with this command using the certificate subject data obtained with the CSR information: 
    
    ```bash
    openssl genrsa -out generated.key 2048
    openssl req -new -sha256 -out generated.csr -key generated.key -subj "{SUBJECT}"
    openssl base64 -in generated.csr
    ```

4. Sign the CSR and get a client certificate. 

    To get the CSR signed, use the encoded CSR in this GraphQL mutation:
    
    ```graphql
    mutation {
        result: signCertificateSigningRequest(csr: "{BASE64_ENCODED_CSR}") {
            certificateChain
            caCertificate
            clientCertificate
        }
    }
    ```
   
    Send the modified GraphQL mutation to the Connector URL including the one-time token fetched with the configuration in the `connector-token` header.

    The response contains a certificate chain, a valid client certificate signed by the Kyma Certificate Authority (CA), and the CA certificate.
    

>     After you receive the certificates, decode the certificate chain with the base64 method and use it in your application. 

Reviewer 2: ```suggestion
    After you receive the certificates, decode the certificate chain with the base64 method and use it in your Application. 
```

Author: The application meant here is not a Kyma Application, that's why it's not capitalized.

>     See also, how to [renew](08-02-maintain-secure-connection-with-compass.md) a client certificate.

Reviewer 2: ```suggestion
>**NOTE:** To learn how to renew a client certificate, read [this](08-02-maintain-secure-connection-with-compass.md) document.
```
